core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2690,7 +2691,14 @@ int complexity() {
     }
 
     @Override public RelDataType getType() {
-      return rexBuilder.typeFactory.leastRestrictive(Util.distinctList(types));
+      if (this.types.isEmpty()) {
+        // Expression is "x IS NULL"
+        return ref.getType();
```

Comment:
It looks strange that `ref.getType()` is considered only in the case of `null-only` sarg.
Should it be included to `this.types` in the constructor instead, so the types are handled in a uniform way?

Reply:
I agree it's a bit strange. It's tricky because the assignment rules for the left-hand type are different from the right-hand types.

The type assignment rules are complicated (e.g. when you compare a REAL to a DECIMAL, do you convert the REAL to a DECIMAL or vice versa?; and hundred other things.  I decided to not spend effort addressing it at this point.  If you can find a SQL example where it matters, log a bug.